Parallel::PerformAlgorithmCallback fails to compile under mocking framework

Test case:
```c++
namespace TestMutableGlobalCache {
struct CacheTag : db::SimpleTag {
  using type = int;
};

template <typename Metavariables>
struct Component {
  using metavariables = Metavariables;
  using chare_type = ActionTesting::MockArrayChare;
  using array_index = int;

  using phase_dependent_action_list =
      tmpl::list<Parallel::PhaseActions<typename Metavariables::Phase,
                                        Metavariables::Phase::Initialization,
                                        tmpl::list<>>>;
};

struct Metavariables {
  using component_list = tmpl::list<Component<Metavariables>>;

  using mutable_global_cache_tags = tmpl::list<CacheTag>;

  enum class Phase { Initialization, Testing, Exit };
};

SPECTRE_TEST_CASE("Unit.ActionTesting.MutableGlobalCache", "[Unit]") {
  using metavars = Metavariables;
  using component = Component<metavars>;

  ActionTesting::MockRuntimeSystem<metavars> runner{{}, {0}};
  ActionTesting::emplace_array_component<component>(
      &runner, ActionTesting::NodeId{0}, ActionTesting::LocalCoreId{0}, 0);

  auto& cache = ActionTesting::cache<component>(runner, 0);
  const auto& element_proxy =
      ::Parallel::get_parallel_component<component>(cache)[0];

  CHECK(Parallel::mutable_cache_item_is_ready<CacheTag>(
      cache, [](const int /*value*/) noexcept {
        return std::unique_ptr<Parallel::Callback>{};
      }));

  CHECK(not Parallel::mutable_cache_item_is_ready<CacheTag>(
      cache, [&element_proxy](const int /*value*/) noexcept {
        return std::unique_ptr<Parallel::Callback>(
            new Parallel::PerformAlgorithmCallback(element_proxy));
      }));
}
}  // namespace TestMutableGlobalCache                                                                         
```
The last statement fails to compile with
```
/home/wthrowe/spectre/src/Parallel/Callback.hpp:80:46: error: invalid operands to binary expression ('PUP::er' and 'ActionTesting::ActionTesting_detail::MockDistributedObjectProxy<(anonymous namespace)::TestMutableGlobalCache::Component<(anonymous namespace)::TestMutableGlobalCache::Metavariables>, brigand::list<>>')
  void pup(PUP::er& p) noexcept override { p | proxy_; }
                                           ~ ^ ~~~~~~
```

--- a/tests/Unit/Framework/ActionTesting.hpp
+++ b/tests/Unit/Framework/ActionTesting.hpp
@@ -483,6 +483,16 @@ class MockDistributedObjectProxy {
                : nullptr;
   }
 
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& /*p*/) noexcept {
+    ERROR(
+        "Should not try to serialize the MockDistributedObjectProxy. "
+        "If you encountered this error you are using the mocking framework "
+        "in a way that it was not intended to be used. It may be possible "
+        "to extend it to more use cases but it is recommended you file an "
+        "issue to discuss before modifying the mocking framework.");
+  }
+
  private:
   // mock_node_ and mock_local_core_ are the (mocked) node and core
   // that this MockDistributedObjectProxy lives on.  This is different
@@ -621,8 +631,8 @@ class MockCollectionOfDistributedObjectsProxy {
                   });
   }
 
-  // clang-tidy: no non-const references
-  void pup(PUP::er& /*p*/) noexcept {  // NOLINT
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& /*p*/) noexcept {
     ERROR(
         "Should not try to serialize the CollectionOfMockDistributedObjects. "
         "If you encountered this error you are using the mocking framework "

--- a/tests/Unit/Framework/ActionTesting.hpp
+++ b/tests/Unit/Framework/ActionTesting.hpp
@@ -523,13 +523,19 @@ class MockCollectionOfDistributedObjectsProxy {
   template <typename InboxTag, typename Data>
   void receive_data(const typename InboxTag::temporal_id& id, const Data& data,
                     const bool enable_if_disabled = false) {
-    // Create (and call) a proxy on the local node and core that references
-    // each of the mock_distributed_objects.
+    // Call (and possibly create/store) a proxy on the local node and core
+    // that references each of the mock_distributed_objects.
     for (const auto& key_value_pair : *mock_distributed_objects_) {
-      MockDistributedObjectProxy<Component, InboxTagList>(
-          mock_node_, mock_local_core_,
-          mock_distributed_objects_->at(key_value_pair.first),
-          inboxes_->operator[](key_value_pair.first))
+      if (proxies_to_mock_distributed_objects_.count(key_value_pair.first) ==
+          0) {
+        proxies_to_mock_distributed_objects_.emplace(std::make_pair(
+            key_value_pair.first,
+            MockDistributedObjectProxy<Component, InboxTagList>(
+                mock_node_, mock_local_core_,
+                mock_distributed_objects_->at(key_value_pair.first),
+                inboxes_->operator[](key_value_pair.first))));
+      }
+      proxies_to_mock_distributed_objects_.at(key_value_pair.first)
           .template receive_data<InboxTag>(id, data, enable_if_disabled);
     }
   }
@@ -544,7 +550,7 @@ class MockCollectionOfDistributedObjectsProxy {
     mock_global_core_ = mock_global_core;
   }
 
-  MockDistributedObjectProxy<Component, InboxTagList> operator[](
+  MockDistributedObjectProxy<Component, InboxTagList>& operator[](
       const Index& index) {
     ASSERT(mock_distributed_objects_->count(index) == 1,
            "Should have exactly one mock distributed object with key '"
@@ -554,9 +560,14 @@ class MockCollectionOfDistributedObjectsProxy {
                << ". Did you forget to add a mock distributed object when "
                   "constructing "
                   "the MockRuntimeSystem?");
-    return MockDistributedObjectProxy<Component, InboxTagList>(
-        mock_node_, mock_local_core_, mock_distributed_objects_->at(index),
-        inboxes_->operator[](index));
+    if (proxies_to_mock_distributed_objects_.count(index) == 0) {
+      proxies_to_mock_distributed_objects_.emplace(std::make_pair(
+          index, MockDistributedObjectProxy<Component, InboxTagList>(
+                     mock_node_, mock_local_core_,
+                     mock_distributed_objects_->at(index),
+                     inboxes_->operator[](index))));
+    }
+    return proxies_to_mock_distributed_objects_.at(index);
   }
 
   // ckLocalBranch should never be called on an array or singleton
@@ -643,6 +654,8 @@ class MockCollectionOfDistributedObjectsProxy {
 
  private:
   CollectionOfMockDistributedObjects* mock_distributed_objects_;
+  std::unordered_map<Index, MockDistributedObjectProxy<Component, InboxTagList>>
+      proxies_to_mock_distributed_objects_;
   Inboxes* inboxes_;
   // mock_node_, mock_local_core_, and mock_global_core_ are the
   // (mock) node and core that the

--- a/tests/Unit/Framework/Tests/Test_ActionTesting.cpp
+++ b/tests/Unit/Framework/Tests/Test_ActionTesting.cpp
@@ -949,4 +949,128 @@ SPECTRE_TEST_CASE("Unit.ActionTesting.NodesAndCores", "[Unit]") {
 }
 
 }  // namespace TestNodesAndCores
+
+namespace TestMutableGlobalCache {
+struct CacheTag : db::SimpleTag {
+  using type = int;
+};
+
+template <int Value>
+struct CacheTagUpdater {
+  static void apply(gsl::not_null<typename CacheTag::type*> tag_value){
+    *tag_value = Value;
+  }
+};
+
+template <typename Metavariables>
+struct Component {
+  using metavariables = Metavariables;
+  using chare_type = ActionTesting::MockArrayChare;
+  using array_index = int;
+
+  using phase_dependent_action_list =
+      tmpl::list<Parallel::PhaseActions<typename Metavariables::Phase,
+                                        Metavariables::Phase::Initialization,
+                                        tmpl::list<>>>;
+};
+
+// This Action does nothing other than set a bool so that we can
+// check if it was called.
+static bool simple_action_to_test_was_called = false;
+struct SimpleActionToTest {
+  template <typename ParallelComponent, typename DbTagsList,
+            typename Metavariables, typename ArrayIndex>
+  static void apply(const db::DataBox<DbTagsList>& /*box*/,
+                    const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/) noexcept {
+    simple_action_to_test_was_called = true;
+  }
+};
+
+struct Metavariables {
+  using component_list = tmpl::list<Component<Metavariables>>;
+
+  using mutable_global_cache_tags = tmpl::list<CacheTag>;
+
+  enum class Phase { Initialization, Testing, Exit };
+};
+
+SPECTRE_TEST_CASE("Unit.ActionTesting.MutableGlobalCache", "[Unit]") {
+  using metavars = Metavariables;
+  using component = Component<metavars>;
+
+  ActionTesting::MockRuntimeSystem<metavars> runner{{}, {0}};
+  ActionTesting::emplace_array_component<component>(
+      &runner, ActionTesting::NodeId{0}, ActionTesting::LocalCoreId{0}, 0);
+
+  auto& cache = ActionTesting::cache<component>(runner, 0);
+  auto& element_proxy = ::Parallel::get_parallel_component<component>(cache)[0];
+
+  CHECK(Parallel::mutable_cache_item_is_ready<CacheTag>(
+      cache, [](const int /*value*/) noexcept {
+        return std::unique_ptr<Parallel::Callback>{};
+      }));
+
+  CHECK(not Parallel::mutable_cache_item_is_ready<CacheTag>(
+      cache, [&element_proxy](const int /*value*/) noexcept {
+        return std::unique_ptr<Parallel::Callback>(
+            new Parallel::PerformAlgorithmCallback<decltype(element_proxy)>(
+                element_proxy));
+      }));
+
+  CHECK(not Parallel::mutable_cache_item_is_ready<CacheTag>(
+      cache, [&element_proxy](const int /*value*/) noexcept {
+        return std::unique_ptr<Parallel::Callback>(
+            new Parallel::SimpleActionCallback<SimpleActionToTest,
+                                               decltype(element_proxy)>(
+                element_proxy));
+      }));
+
+  // Should be no queued simple actions.
+  CHECK(ActionTesting::is_simple_action_queue_empty<component>(runner, 0));
+
+  // After we mutate the item, then SimpleActionToTest should be queued...
+  Parallel::mutate<CacheTag, CacheTagUpdater<1>>(cache);
+  CHECK(ActionTesting::number_of_queued_simple_actions<component>(runner, 0) ==
+        1);
+  // ... so invoke it
+  ActionTesting::invoke_queued_simple_action<component>(make_not_null(&runner),
+                                                        0);
+  // ... and test that the action was called.
+  CHECK(simple_action_to_test_was_called);
+
+  // Now reset for another call.
+  simple_action_to_test_was_called = false;
+
+  // Simple actions can be called on entire components
+  // (i.e. all elements at once as opposed to one element at a time).
+  auto& all_elements_proxy =
+      ::Parallel::get_parallel_component<component>(cache);
+  CHECK(not Parallel::mutable_cache_item_is_ready<CacheTag>(
+      cache, [&all_elements_proxy](const int /*value*/) noexcept {
+        return std::unique_ptr<Parallel::Callback>(
+            new Parallel::SimpleActionCallback<SimpleActionToTest,
+                                               decltype(all_elements_proxy)>(
+                all_elements_proxy));
+      }));
+
+  // Should be no queued simple actions.
+  CHECK(ActionTesting::is_simple_action_queue_empty<component>(runner, 0));
+
+  // After we mutate the item, then SimpleActionToTest should be queued...
+  Parallel::mutate<CacheTag, CacheTagUpdater<3>>(cache);
+  CHECK(ActionTesting::number_of_queued_simple_actions<component>(runner, 0) ==
+        1);
+  // ... so invoke it
+  ActionTesting::invoke_queued_simple_action<component>(make_not_null(&runner),
+                                                        0);
+  // ... and test that the action was called.
+  CHECK(simple_action_to_test_was_called);
+
+  // Currently perform_algorithm cannot be called on all elements at
+  // once (in the ActionTesting framework), but this is not difficult
+  // to add in the future.
+}
+}  // namespace TestMutableGlobalCache
+
 }  // namespace

